src/object/srv_obj.c

```diff
@@ -782,45 +782,45 @@ obj_local_rw(crt_rpc_t *rpc, struct ds_cont_hdl *cont_hdl,
 	       orw->orw_bulks.ca_count != 0);
 	oca = daos_oclass_attr_find(orw->orw_oid.id_pub);
 
-	/* Prepare IO descriptor */
-	if (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE ||
-	    opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_TGT_UPDATE) {
-		bulk_op = CRT_BULK_GET;
-		if (oca->ca_resil == DAOS_RES_EC) {
-			unsigned int	tgt_idx = orw->orw_oid.id_shard -
-							orw->orw_start_shard;
+	if (oca->ca_resil == DAOS_RES_EC) {
+		unsigned int	tgt_idx = orw->orw_oid.id_shard -
+					  orw->orw_start_shard;
 
-			D_ALLOC_ARRAY(skip_list, orw->orw_nr);
-			if (skip_list == NULL) {
-				D_GOTO(out, rc = -DER_NOMEM);
-			}
-			if (tgt_idx >= oca->u.ec.e_p) {
-				rc = ec_data_target(tgt_idx - oca->u.ec.e_p,
-						    orw->orw_nr, tmp_iods,
-						    oca, skip_list);
-			} else {
-				if (tgt_idx == 0) {
-					rc = ec_copy_iods(tmp_iods,
-							    orw->orw_nr,
-							    &cpy_iods);
-					if (rc) {
-						D_ERROR(
-						DF_UOID"EC update failed: %d\n",
-						DP_UOID(orw->orw_oid), rc);
-						goto out;
-					}
-					tmp_iods = cpy_iods;
+		D_ALLOC_ARRAY(skip_list, orw->orw_nr);
+		if (skip_list == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+		if (tgt_idx >= oca->u.ec.e_p) {
+			rc = ec_data_target(tgt_idx - oca->u.ec.e_p,
+					    orw->orw_nr, tmp_iods,
+					    oca, skip_list);
+		} else {
+			if (tgt_idx == 0) {
+				rc = ec_copy_iods(tmp_iods,
+						    orw->orw_nr,
+						    &cpy_iods);
+				if (rc) {
+					D_ERROR(
+					DF_UOID"EC update failed: %d\n",
+					DP_UOID(orw->orw_oid), rc);
+					goto out;
```

Comment:
Do you need free skip_list in these failure cases?

Reply:
it is freed in below L882 ~ L883, agreed that some detailed code can be refined by following patch.